### DISCUSSION TITLE
Substructure Search and aromaticity

### DISCUSSION THREAD

**Discussion by kienerj** [rdkit/rdkit#5313] [Q&A]
A topic that keeps coming up for me especially in comparison to other tools. A recent example:

```
query = Chem.MolFromSmiles('C12=C(C=CC=C2)NSC1')
molecule = Chem.MolFromSmiles('O=c1s[nH]c2ccccc12')
print(molecule.HasSubstructMatch(query))
False
```

![image](https://user-images.githubusercontent.com/1567309/169524391-6ea379d7-b651-4464-8b8a-21b71285be76.png)

Intuitively I myself and chemists would expect a substructure match here. 
I assume there is not match due to the different aromaticity and the left side ring:

![image](https://user-images.githubusercontent.com/1567309/169525604-ebf8e76f-9eba-46a6-8e51-7c95d45fb625.png)

I'm not even sure why the 5-ring in the full molecule is aromatic, but never mind.
The core question here is to get rdkit and/or the rdkit cartridge to return a match for these structures given the above smiles input. Are there any flags to set?

I have now also found below possibly relevant gist form Paolo:
https://gist.github.com/ptosco/e410e45278b94e8f047ff224193d7788
But it seems not very generally applicable but more geared to the specific use-case. In many cases you would not want the generic bonds so it can't be a rule/flag that you always apply for every single search but that would be needed.

How can we properly handle cases like this to get expected results?

**Comment by ptosco** [comment]
@kienerj Hi Joos, you may set a different aromaticity model. e.g.:
```
query = Chem.MolFromSmiles('C12=C(C=CC=C2)NSC1')
molecule = Chem.MolFromSmiles('O=c1s[nH]c2ccccc12')

Chem.Kekulize(molecule, True)
Chem.Kekulize(query, True)

Chem.SetAromaticity(molecule, Chem.AromaticityModel.AROMATICITY_MDL)
Chem.SetAromaticity(query, Chem.AromaticityModel.AROMATICITY_MDL)

molecule
```
![image](https://user-images.githubusercontent.com/5244385/169542736-ac48d670-5c9d-4854-ad25-8ba9e25a0b42.png)
```
query
```
![image](https://user-images.githubusercontent.com/5244385/169542821-476e9c2f-ea14-4500-9516-4bb282e0108a.png)
```
molecule.HasSubstructMatch(query)
True
```

**Comment by kienerj** [comment]
Thanks Paolo, What if we assume "molecule" is in the database (rdkit cartridge) and "query" is the query? Would I have to apply above aromatically model before saving the molecule? But that would require me to save the molecule object/pickle directly and not via smiles/molfile?

**Reply by greglandrum** [reply]
@kienerj, there's currently no way to change the aromaticity model used in the database.
Even if you were able to do so, you would still end up running into aromaticity edge cases; they would just be different edge cases. 

My general suggestion when situations like this pop up is to just explain the situation to the users. This situation is exactly analogous to why cyclopentane:
![image](https://user-images.githubusercontent.com/540511/169752929-1e785275-d9c9-49f3-bb72-0de6043783f2.png)
is not considered to be a substructure of fluorene:
![image](https://user-images.githubusercontent.com/540511/169753001-cdfd7b5d-17c4-4c62-9251-6615c5e08330.png)

As for why the original molecule is aromatic, the RDKit's aromaticity model is described here: https://www.rdkit.org/docs/RDKit_Book.html#aromaticity

**Comment by kienerj** [comment]
Thinking out loud it might be a good idea to have an additional option in the AdjustQueryParameters that converts a bond to a query bond of type single or aromatic or double or aromatic depending on what the initial bond is. This would make it possible to hit the substructure if it is aromatic or not. Converting to generic bond seems too general and lead to confusing hits (a single bond will match a double bond or vice versa).
It would also make sense that this is applied to non-ring bonds of the query structure of one wishes to do so and get a match for a query like below:

![image](https://user-images.githubusercontent.com/1567309/170192674-d862013b-4aef-470d-b3d7-8a6e5676bddf.png)

I actually wanted to try this out manually but it seems it is not possible to create query bonds from python?

![image](https://user-images.githubusercontent.com/1567309/170192999-f896a1fb-40a0-42f2-a974-f21908a3a57d.png)

Is the only way to get a query bond by already importing the molecule into rdkit that way?

**Reply by greglandrum** [reply]
The easiest way to get a `QueryBond` in python is to use `Chem.BondFromSmarts()`:
```
In [2]: b = Chem.BondFromSmarts("-,:,=")

In [3]: b
Out[3]: <rdkit.Chem.rdchem.QueryBond at 0x7fb7af7e9ee0>
```

**Reply by kienerj** [reply]
Thanks Greg.  With that I can achieve my idea (up for decision if it's a good one) with below code but probably further testing needed.

```
query = Chem.MolFromSmiles('C1=CC(NSC2)=C2C=C1')
molecule = Chem.MolFromSmiles('O=c1s[nH]c2ccccc12')
qmol = Chem.RWMol(query)

for bond in qmol.GetBonds():
    if bond.IsInRing() and not bond.GetIsAromatic():
        if bond.GetBondType() == Chem.BondType.SINGLE:
            qb = Chem.BondFromSmarts("-,:")
        elif bond.GetBondType() == Chem.BondType.DOUBLE:
            qb = Chem.BondFromSmarts("=,:")
        qmol.ReplaceBond(bond.GetIdx(), qb)
# qmol must not be sanitized or else bond types get changed again
qmol
```
![image](https://user-images.githubusercontent.com/1567309/170226001-2380395f-0ae8-4345-a2de-15739c84700f.png)

**Reply by greglandrum** [reply]
I personally wouldn't do this (as I said before, there are always going to be edge cases and you might as well teach users to know that and to draw their queries appropriately), but if you think this is the right thing for your users, I don't see anything technically wrong with it.

You could make it marginally more efficient like this:
```
singleOrAromatic = Chem.BondFromSmarts("-,:")
doubleOrAromatic = Chem.BondFromSmarts("=,:")
for bond in qmol.GetBonds():
    if bond.IsInRing() and not bond.GetIsAromatic():
        if bond.GetBondType() == Chem.BondType.SINGLE:
            qmol.ReplaceBond(bond.GetIdx(), singleOrAromatic)
       elif bond.GetBondType() == Chem.BondType.DOUBLE:
            qmol.ReplaceBond(bond.GetIdx(), doubleOrAromatic)
```
